# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/public/CompileOptions.h
# Commit: 0174baa33cf7
# Full Hash: 0174baa33cf7cd6fd0db96c4f1411e57eccd2bf0
# Author: Debadree Chatterjee <debadree333@gmail.com>
# Date: 2024-11-03 21:45:44
# Regressor Bug: 1927195
# File Overlap Count: 1
# Description:
#   Bug 1927195 - Part 4: Prevent parsing using & await using syntax based on pref. r=arai
#   
#   Depends on D226993
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D226994
# ==============================================================================

diff -r c0cc438b2397 -r 0174baa33cf7 js/public/CompileOptions.h
--- a/js/public/CompileOptions.h	Sun Nov 03 08:32:26 2024 +0000
+++ b/js/public/CompileOptions.h	Sun Nov 03 08:32:27 2024 +0000
@@ -62,7 +62,10 @@
 
 #include "js/CharacterEncoding.h"  // JS::ConstUTF8CharsZ
 #include "js/ColumnNumber.h"       // JS::ColumnNumberOneOrigin
-#include "js/TypeDecls.h"          // JS::MutableHandle (fwd)
+#ifdef ENABLE_EXPLICIT_RESOURCE_MANAGEMENT
+#  include "js/Prefs.h"  // JS::Prefs::*
+#endif
+#include "js/TypeDecls.h"  // JS::MutableHandle (fwd)
 
 namespace js {
 class FrontendContext;
@@ -125,7 +128,12 @@
   PrefableCompileOptions()
       : importAttributes_(false),
         sourcePragmas_(true),
-        throwOnAsmJSValidationFailure_(false) {}
+#ifdef ENABLE_EXPLICIT_RESOURCE_MANAGEMENT
+        explicitResourceManagement_(
+            JS::Prefs::experimental_explicit_resource_management()),
+#endif
+        throwOnAsmJSValidationFailure_(false) {
+  }
 
   bool importAttributes() const { return importAttributes_; }
   PrefableCompileOptions& setImportAttributes(bool enabled) {
@@ -133,6 +141,16 @@
     return *this;
   }
 
+#ifdef ENABLE_EXPLICIT_RESOURCE_MANAGEMENT
+  bool explicitResourceManagement() const {
+    return explicitResourceManagement_;
+  }
+  PrefableCompileOptions& setExplicitResourceManagement(bool enabled) {
+    explicitResourceManagement_ = enabled;
+    return *this;
+  }
+#endif
+
   // Enable/disable support for parsing '//(#@) source(Mapping)?URL=' pragmas.
   bool sourcePragmas() const { return sourcePragmas_; }
   PrefableCompileOptions& setSourcePragmas(bool flag) {
@@ -170,6 +188,9 @@
     PrintFields_(importAttributes_);
     PrintFields_(sourcePragmas_);
     PrintFields_(throwOnAsmJSValidationFailure_);
+#  ifdef ENABLE_EXPLICIT_RESOURCE_MANAGEMENT
+    PrintFields_(explicitResourceManagement_);
+#  endif
 #  undef PrintFields_
 
     switch (asmJSOption_) {
@@ -199,6 +220,12 @@
   // The context has specified that source pragmas should be parsed.
   bool sourcePragmas_ : 1;
 
+#ifdef ENABLE_EXPLICIT_RESOURCE_MANAGEMENT
+  // The context has specified that explicit resource management syntax
+  // should be parsed.
+  bool explicitResourceManagement_ : 1;
+#endif
+
   // ==== asm.js options. ====
   bool throwOnAsmJSValidationFailure_ : 1;
 
@@ -381,6 +408,11 @@
 
   bool importAttributes() const { return prefableOptions_.importAttributes(); }
   bool sourcePragmas() const { return prefableOptions_.sourcePragmas(); }
+#ifdef ENABLE_EXPLICIT_RESOURCE_MANAGEMENT
+  bool explicitResourceManagement() const {
+    return prefableOptions_.explicitResourceManagement();
+  }
+#endif
   bool throwOnAsmJSValidationFailure() const {
     return prefableOptions_.throwOnAsmJSValidationFailure();
   }